_posts/2025-01-08-working-with-graph-databases-in-python.md

```diff
@@ -0,0 +1,278 @@
+---
+title:  "Working with RDF Graph Databases in Python"
+author: Leo Rudczenko
+categories:
+  - Software Engineering
+tags:
+  - python
+  - graph
+  - database
+  - rdf
+  - linked-data
+---
+
+## What are Graph Databases?
+
+The logic behind a [Graph Database](https://en.wikipedia.org/wiki/Graph_database),
+comes from graph theory in mathematics, where a network of nodes representing data are connected using edges.
+This makes a graph database more flexible than a typical hierarchical database,
+as new nodes can be added and connected to any existing node, often without constraints.
+This lack of constraints can be convenient, but it does mean that a user will need
+additional context regarding the existing nodes before adding new ones, as the existing
+data is unpredictable.
+Constraints can be added to a graph database using additional tools if required.
+For example, when using
+a [Resource Description Framework (RDF)](https://en.wikipedia.org/wiki/Resource_Description_Framework)
+graph, you can add constraints
+using [Shapes Constraint Language (SHACL)](https://en.wikipedia.org/wiki/SHACL).
+
+In a graph database, everything is represented using URIs (Uniform Resource Identifier) or literals.
+The URIs are what allows us to create new connections between nodes in the graph,
+and even connections between multiple graphs.
+For this to work, data is stored in `triples`. In a triple, we have 3 pieces of information:
+- Subject: the row in a typical database, referring to one feature (URI)
+- Predicate: the column in a typical database, referring to one attribute name (URI)
+- Object: the value found at the given row/column, the attribute value (URI or literal)
+
+The predicate URI should also be a subject itself, containing more triples that describe its definition.
+This means that all of the context behind the data is described within the graph database,
+an advantage which cannot be done in a relational database.
+
+Let's apply this concept to a lithology example. If we want to represent Rhyolite's lithology classification
```

Comment:
Maybe add some explanation for non-geological audience:  "Let's apply this concept to an example using rock classification data. If we want to describe what class of rock Rhyolite ( type of igneous rock) is using a triple, we could do the following: 
- Subject: URI representing the rock class `Rhyolite`
- Predicate: URI representing the relationship child-to-parent, or class-to-superclass
- Object: URI representing the rock class `Igneous Rock`